jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
Sorry that it was an offline discussion. We want the tool is flexible enough so that when people (they are advanced user) decide to provide their own custom transform function, the function can handle the most cases (including the multi-js-args => 1-native-arg )

For primary users, they just use the builtin transform function, and don't need to touch the iterator.

One possible example in graphic related case
```
//the js api
drawLine(x1, y1, x2, y2);
drawPoint(x1, y1);
drawTriangle(x1, y1, x2, y2. x3, y3);

//in native resouce (API provided by firmware)
typedef  struct { int x; int y; } GPoint;
void LCD_draw_point (GPoint x);
void LCD_draw_line (GPoint x1, GPoint x2);
void LCD_draw_triangle (GPoint x1, GPoint x2, GPoint x3 );

// In the binding layer, although we can create a empty GPoint obj, 
// and assign one js_arg to one member of the obj. 
// But in this way, the mapping array is tedious, and cost more code space
// Since all graphic api needs the struct GPoint as input
// we can provide a custom transform function to convert 2 js arg into one GPoint obj

jerry_value_t my_custom_gpoint_transform (iterator_p)
{
   jerry_value_t x = jerryx_arg_iterator_pop (iterator_p);
   jerry_value_t to_number = jerry_value_to_number (x);
   Gpoint*dest = (Gpoint*)jerryx_arg_iterator_get_dest (iterator_p);
   dest->x = jerry_get_number_value (to_number);
   jerry_value_t x = jerryx_arg_iterator_pop (iterator_p);
   to_number = jerry_value_to_number (y);
   dest->y = jerry_get_number_value (to_number);

   jerry_release_value (to_number);
   return jerry_create_undefined (); // I omit the error check code in the function.
} 

jerry_value_t drawTrangle_handle(...)
{
   GPoint x1, x2, x3;
   jerryx_arg_t mapping = 
  [ 
    jerryx_arg_ignore (), // ignore this
    jerryx_arg_custom(&x1, 0, my_custom_gpoint_transform ), // the first point
    jerryx_arg_custom(&x2, 0, my_custom_gpoint_transform ),//the second point
    jerryx_arg_custom(&x3, 0, my_custom_gpoint_transform ), //the third point
  ]
...
}
```